### PR TITLE
Code for the Robot field position from april tag

### PR DESCRIPTION
**Pull request by SAKETH11111**
In this code, we implemented a Vision subsystem using PhotonVision, with a PhotonCamera and PhotonPoseEstimator, to periodically update and log the estimated global pose of the robot.

### CODE REVIEW

File: src/main/java/frc/robot/Robot.java [hunk 1]
```diff
@@ -13,14 +13,26 @@
 
 package frc.robot;
 
+import edu.wpi.first.apriltag.AprilTagFieldLayout;
```
Comment (CanineSatsuma6) [inline src/main/java/frc/robot/Robot.java]: We can revert all of the changes in this file. As is, we'll constantly be scanning for April Tags when the robot starts up, but we disabled the command scheduler, so none of our other control logic will actually execute

File: src/main/java/frc/robot/subsystems/vision/Vision.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    private final PhotonCamera camera;
+    private final PhotonPoseEstimator poseEstimator;
+    private final VisionIOInputsAutoLogged inputs = new VisionIOInputsAutoLogged();
+    private Pose2d memory = new Pose2d();
+
+    public Vision(VisionIO io) {
+        this.io = io;
+        camera = new PhotonCamera("frontCam");
+        poseEstimator = new PhotonPoseEstimator(
+                /* AprilTagFieldLayout */ null,
```
Comment (CanineSatsuma6) [inline src/main/java/frc/robot/subsystems/vision/Vision.java]: We'll want to load the `AprilTagFieldLayout` object in the same way you're currently doing in Robot.java right now

File: src/main/java/frc/robot/subsystems/vision/VisionIOPhotonlib.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  public synchronized void updateInputs(VisionIOInputs inputs) {
+    inputs.captureTimestamp = captureTimestamp;
+    inputs.cornerX = cornerX;
+    inputs.cornerY = cornerY;
+  }
+
+  @Override
+  public void setLeds(boolean enabled) {
+    camera.setLED(enabled ? VisionLEDMode.kOn : VisionLEDMode.kOff);
+  }
```
Comment (CanineSatsuma6) [inline src/main/java/frc/robot/subsystems/vision/VisionIOPhotonlib.java]: We don't have LEDs on our Pi/camera, so we can get rid of this method.

File: src/main/java/frc/robot/subsystems/vision/VisionIO.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  }
+
+  /** Updates the set of loggable inputs. */
+  public default void updateInputs(VisionIOInputs inputs) {}
+
+  /** Enabled or disabled vision LEDs. */
+  public default void setLeds(boolean enabled) {}
+
+  /** Sets the pipeline number. */
+  public default void setPipeline(int pipeline) {}
```
Comment (CanineSatsuma6) [inline src/main/java/frc/robot/subsystems/vision/VisionIO.java]: We don't have LEDs, and we'll only have one pipeline, so we don't need either of these methods.

File: src/main/java/frc/robot/subsystems/vision/VisionIO.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+/** Vision subsystem hardware interface. */
+public interface VisionIO {
+  /** The set of loggable inputs for the vision subsystem. */
+  @AutoLog  
+  public static class VisionIOInputs {
+    public double captureTimestamp = 0.0;
+    public double[] cornerX = new double[] {};
+    public double[] cornerY = new double[] {};
+    public boolean simpleValid = false;
+    public double simpleAngle = 0.0;
```
Comment (CanineSatsuma6) [inline src/main/java/frc/robot/subsystems/vision/VisionIO.java]: We don't use these inputs, and I'm not sure what they represent, so we can remove these as well

**Review by CanineSatsuma6 — CHANGES_REQUESTED** [review]
A couple small things, but you're really close. Nice job!

**Review by edenreicherts — COMMENTED** [review]
Other than the changes that were previously mentioned, I didn't notice anything to adjust! Great work!

**Comment by SAKETH11111** [comment]
Thanks for the comments, Collin. I have made the changes according to comments and I think it is good to go! 😊

**Review by cedricliu07 — CHANGES_REQUESTED** [review]
Good, formatting is slightly off  though in vision and visionphoton

**Review by cedricliu07 — DISMISSED** [review]
Good Changes :D

File: src/main/java/frc/robot/Constants.java [hunk 6]
```diff
@@ -71,6 +72,18 @@ public static class Drive
         public static final Rotation2d MODULE_BR_OFFSET   = Rotation2d.fromRadians(-1.63);
     }
 
+    public static class Vision
+    {
+        public static final String CAMERA_NAME = "frontCam";
+        // Measurement from the camera to the center of the robot
+        public static final int PARAMETER_X = (int)Units.inchesToMeters(12);
+        // Measurement from the camera to the side of the robot
+        public static final int PARAMETER_Y = (int)Units.inchesToMeters(0);
```
Comment (TheWonderfulWit) [inline src/main/java/frc/robot/Constants.java:81]: The camera is not on the ground shouldn't it be above 0

**Review by TheWonderfulWit — DISMISSED** [review]
I didn't see any problems good job

File: src/main/java/frc/robot/RobotContainer.java [hunk 7]
```diff
@@ -42,6 +45,7 @@
 public class RobotContainer
 {
     // Subsystems
+    private Vision         _vision;
```
Comment (AbbySliwinski) [inline src/main/java/frc/robot/RobotContainer.java]: make final and add a warning suppression statement

File: src/main/java/frc/robot/subsystems/vision/VisionIOPhotonlib.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+            List<Double> cornerYList = new ArrayList<>();
+            for (PhotonTrackedTarget target : result.getTargets())
+            {
+                for (TargetCorner corner : target.getDetectedCorners())
+                {
+                    cornerXList.add(corner.x);
+                    cornerYList.add(corner.y);
+                }
+            }
+            Optional<EstimatedRobotPose> estimatedPose = getEstimatedGlobalPose(_lastPose, result);
```
Comment (AbbySliwinski) [inline src/main/java/frc/robot/subsystems/vision/VisionIOPhotonlib.java:74]: add white space

File: src/main/java/frc/robot/subsystems/vision/VisionIOPhotonlib.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+            Pose2d                       newPose       = new Pose2d();
+            _hasPose = false;
+            if (estimatedPose.isPresent())
```
Comment (AbbySliwinski) [inline src/main/java/frc/robot/subsystems/vision/VisionIOPhotonlib.java:78]: create local variable for has pose and update global variable in synchronized

File: src/main/java/frc/robot/subsystems/vision/VisionIOPhotonlib.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+/** Vision hardware implementation for PhotonVision. */
+public class VisionIOPhotonlib implements VisionIO
+{
+    private final PhotonCamera  _camera           = new PhotonCamera(Constants.Vision.CAMERA_NAME);
+    private double              _captureTimestamp = 0.0;
+    private double[]            _cornerX          = new double[] {};
+    private double[]            _cornerY          = new double[] {};
+    private PhotonPoseEstimator _poseEstimator;
+    private Pose2d              _lastPose;
```
Comment (AbbySliwinski) [inline src/main/java/frc/robot/subsystems/vision/VisionIOPhotonlib.java]: make new pose2D

**Review by AbbySliwinski — CHANGES_REQUESTED** [review]

**Review by AbbySliwinski — APPROVED** [review]

**Review by CanineSatsuma6 — APPROVED** [review]

**Pull request merged**